In [1]:
from google.colab import drive
drive.mount('/content/drive')


Mounted at /content/drive


In [2]:
import numpy as np                                                 # Implemennts milti-dimensional array and matrices
import pandas as pd                                                # For data manipulation and analysis
# import pandas_profiling
import matplotlib.pyplot as plt                                    # Plotting library for Python programming language and it's numerical mathematics extension NumPy
import seaborn as sns                                              # Provides a high level interface for drawing attractive and informative statistical graphics
%matplotlib inline
sns.set()

from subprocess import check_output ###Assignment



In [3]:
# loading of the data
Tesla_df = pd.read_csv("/content/drive/MyDrive/Abhijit/Capstone/Tesla - Deaths.csv")

Tesla_df.head()

,Case #,Year,Date,Country,State,Description,Deaths,Tesla driver,Tesla occupant,Other vehicle,...,Verified Tesla Autopilot Deaths,Verified Tesla Autopilot Deaths + All Deaths Reported to NHTSA SGO,Unnamed: 16,Unnamed: 17,Source,Note,Deceased 1,Deceased 2,Deceased 3,Deceased 4
0,294,2022,1/17/23,USA,CA,Tesla crashes into back of semi,1,1,-,-,...,-,-,https://web.archive.org/web/20221222203930/ht...,https://web.archive.org/web/20221222203930/ht...,https://web.archive.org/web/20230118162813/ht...,NaN,NaN,NaN,NaN,NaN
1,293,2022,1/7/23,Canada,-,Tesla crashes,1,1,-,-,...,-,-,https://web.archive.org/web/20221222203930/ht...,https://web.archive.org/web/20221222203930/ht...,https://web.archive.org/web/20230109041434/ht...,NaN,Taren Singh Lal,NaN,NaN,NaN
2,292,2022,1/7/23,USA,WA,"Tesla hits pole, catches on fire",1,-,1,-,...,-,-,https://web.archive.org/web/20221222203930/ht...,https://web.archive.org/web/20221222203930/ht...,https://web.archive.org/web/20230107232745/ht...,NaN,NaN,NaN,NaN,NaN
3,291,2022,12/22/22,USA,GA,Tesla crashes and burns,1,1,-,-,...,-,-,https://web.archive.org/web/20221222203930/ht...,https://web.archive.org/web/20221222203930/ht...,https://web.archive.org/web/20221222203930/ht...,NaN,NaN,NaN,NaN,NaN
4,290,2022,12/19/22,Canada,-,Tesla crashes into storefront,1,-,-,-,...,-,-,https://web.archive.org/web/20221223203725/ht...,https://web.archive.org/web/20221223203725/ht...,https://web.archive.org/web/20221223203725/ht...,NaN,NaN,NaN,NaN,NaN


# 1. Preliminary data inspection and cleaning

In [4]:
Tesla_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 294 entries, 0 to 293
Data columns (total 24 columns):
 #   Column                                                                Non-Null Count  Dtype  
---  ------                                                                --------------  -----  
 0   Case #                                                                294 non-null    int64  
 1   Year                                                                  294 non-null    int64  
 2   Date                                                                  294 non-null    object 
 3    Country                                                              294 non-null    object 
 4    State                                                                294 non-null    object 
 5    Description                                                          294 non-null    object 
 6    Deaths                                                               294 non-null    int64  
 7  

In [5]:
Tesla_df.shape

(294, 24)

In [6]:
Tesla_df.columns

Index(['Case #', 'Year', 'Date', ' Country ', ' State ', ' Description ',
       ' Deaths ', ' Tesla driver ', ' Tesla occupant ', ' Other vehicle ',
       ' Cyclists/ Peds ', ' TSLA+cycl / peds ', ' Model ',
       ' Autopilot claimed ', ' Verified Tesla Autopilot Deaths ',
       ' Verified Tesla Autopilot Deaths + All Deaths Reported to NHTSA SGO ',
       'Unnamed: 16', 'Unnamed: 17', ' Source ', ' Note ', ' Deceased 1 ',
       ' Deceased 2 ', ' Deceased 3 ', ' Deceased 4 '],
      dtype='object')



* From the data information it is clear that the some data has some missing values
I don't need some of the columns like 'unnamed 16', 'unnamed 17', 'note' and 'source' so we can remove them also the deceased counts are very less with respect to the entries of the accidents. Therefore, these columns are not at all useful for analysis. Basically we can remove them.

In [7]:
# Data cleaning by column removal
Tesla_df_clean = Tesla_df.drop(['Unnamed: 16', 'Unnamed: 17', ' Source ', ' Note ', ' Deceased 1 ',
       ' Deceased 2 ', ' Deceased 3 ', ' Deceased 4 '], axis=1)

In [8]:
Tesla_df_clean.head()

,Case #,Year,Date,Country,State,Description,Deaths,Tesla driver,Tesla occupant,Other vehicle,Cyclists/ Peds,TSLA+cycl / peds,Model,Autopilot claimed,Verified Tesla Autopilot Deaths,Verified Tesla Autopilot Deaths + All Deaths Reported to NHTSA SGO
0,294,2022,1/17/23,USA,CA,Tesla crashes into back of semi,1,1,-,-,-,1,-,-,-,-
1,293,2022,1/7/23,Canada,-,Tesla crashes,1,1,-,-,-,1,-,-,-,-
2,292,2022,1/7/23,USA,WA,"Tesla hits pole, catches on fire",1,-,1,-,-,1,-,-,-,-
3,291,2022,12/22/22,USA,GA,Tesla crashes and burns,1,1,-,-,-,1,-,-,-,-
4,290,2022,12/19/22,Canada,-,Tesla crashes into storefront,1,-,-,-,1,1,-,-,-,-


In [9]:
Tesla_df_clean.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 294 entries, 0 to 293
Data columns (total 16 columns):
 #   Column                                                                Non-Null Count  Dtype 
---  ------                                                                --------------  ----- 
 0   Case #                                                                294 non-null    int64 
 1   Year                                                                  294 non-null    int64 
 2   Date                                                                  294 non-null    object
 3    Country                                                              294 non-null    object
 4    State                                                                294 non-null    object
 5    Description                                                          294 non-null    object
 6    Deaths                                                               294 non-null    int64 
 7    Tesla d

In [10]:
Tesla_df_clean.columns

Index(['Case #', 'Year', 'Date', ' Country ', ' State ', ' Description ',
       ' Deaths ', ' Tesla driver ', ' Tesla occupant ', ' Other vehicle ',
       ' Cyclists/ Peds ', ' TSLA+cycl / peds ', ' Model ',
       ' Autopilot claimed ', ' Verified Tesla Autopilot Deaths ',
       ' Verified Tesla Autopilot Deaths + All Deaths Reported to NHTSA SGO '],
      dtype='object')

In [11]:
Tesla_df_clean.columns = Tesla_df_clean.columns.str.strip() # To strip the space in the ccolumn names


In [12]:
Tesla_df_clean.columns

Index(['Case #', 'Year', 'Date', 'Country', 'State', 'Description', 'Deaths',
       'Tesla driver', 'Tesla occupant', 'Other vehicle', 'Cyclists/ Peds',
       'TSLA+cycl / peds', 'Model', 'Autopilot claimed',
       'Verified Tesla Autopilot Deaths',
       'Verified Tesla Autopilot Deaths + All Deaths Reported to NHTSA SGO'],
      dtype='object')

In [13]:
# Check of the records where tesla occupants were present
Tesla_df_clean['Tesla occupant'].value_counts()

,count
Tesla occupant,
-,242
1,39
2,3
3,1


In [14]:
Tesla_df_clean['Tesla driver'].unique()

array(['1', ' - ', nan], dtype=object)

In [15]:
# Checking of the missing value percentage in the cleaned data
Tesla_df_clean.isnull().sum()/len(Tesla_df_clean)*100

,0
Case #,0.000000
Year,0.000000
Date,0.000000
Country,0.000000
State,0.000000
Description,0.000000
Deaths,0.000000
Tesla driver,1.700680
Tesla occupant,3.061224
Other vehicle,1.360544


In [16]:
Tesla_df_clean['Autopilot claimed'].unique()

array([' - ', '1', nan, '2'], dtype=object)

In [17]:
Tesla_df_clean.fillna(' - ', inplace=True)

In [18]:
Tesla_df_clean['Autopilot claimed'].unique()

array([' - ', '1', '2'], dtype=object)

In [19]:
# Checking of the missing value percentage after replacing the nan with " - " the cleaned data
Tesla_df_clean.isnull().sum()/len(Tesla_df_clean)*100

,0
Case #,0.0
Year,0.0
Date,0.0
Country,0.0
State,0.0
Description,0.0
Deaths,0.0
Tesla driver,0.0
Tesla occupant,0.0
Other vehicle,0.0


In [20]:
Tesla_df_clean.head()

,Case #,Year,Date,Country,State,Description,Deaths,Tesla driver,Tesla occupant,Other vehicle,Cyclists/ Peds,TSLA+cycl / peds,Model,Autopilot claimed,Verified Tesla Autopilot Deaths,Verified Tesla Autopilot Deaths + All Deaths Reported to NHTSA SGO
0,294,2022,1/17/23,USA,CA,Tesla crashes into back of semi,1,1,-,-,-,1,-,-,-,-
1,293,2022,1/7/23,Canada,-,Tesla crashes,1,1,-,-,-,1,-,-,-,-
2,292,2022,1/7/23,USA,WA,"Tesla hits pole, catches on fire",1,-,1,-,-,1,-,-,-,-
3,291,2022,12/22/22,USA,GA,Tesla crashes and burns,1,1,-,-,-,1,-,-,-,-
4,290,2022,12/19/22,Canada,-,Tesla crashes into storefront,1,-,-,-,1,1,-,-,-,-


In [21]:
Tesla_df_clean.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 294 entries, 0 to 293
Data columns (total 16 columns):
 #   Column                                                              Non-Null Count  Dtype 
---  ------                                                              --------------  ----- 
 0   Case #                                                              294 non-null    int64 
 1   Year                                                                294 non-null    int64 
 2   Date                                                                294 non-null    object
 3   Country                                                             294 non-null    object
 4   State                                                               294 non-null    object
 5   Description                                                         294 non-null    object
 6   Deaths                                                              294 non-null    int64 
 7   Tesla driver              

In [22]:
Tesla_df_clean['Date'] = pd.to_datetime(Tesla_df_clean['Date'], format='%m/%d/%y', errors='coerce') # pandas date time format

In [23]:
# dropping year coloum as Date column already has year
Tesla_df_clean = Tesla_df_clean.drop(['Year'], axis=1)

In [24]:
Tesla_df_clean.head()

,Case #,Date,Country,State,Description,Deaths,Tesla driver,Tesla occupant,Other vehicle,Cyclists/ Peds,TSLA+cycl / peds,Model,Autopilot claimed,Verified Tesla Autopilot Deaths,Verified Tesla Autopilot Deaths + All Deaths Reported to NHTSA SGO
0,294,2023-01-17,USA,CA,Tesla crashes into back of semi,1,1,-,-,-,1,-,-,-,-
1,293,2023-01-07,Canada,-,Tesla crashes,1,1,-,-,-,1,-,-,-,-
2,292,2023-01-07,USA,WA,"Tesla hits pole, catches on fire",1,-,1,-,-,1,-,-,-,-
3,291,2022-12-22,USA,GA,Tesla crashes and burns,1,1,-,-,-,1,-,-,-,-
4,290,2022-12-19,Canada,-,Tesla crashes into storefront,1,-,-,-,1,1,-,-,-,-


# 2. Exploratory Data Analysis

In [25]:
# year wise accident analysis
Tesla_df_clean['Date'].dt.year.value_counts()

,count
Date,
2022,93
2021,58
2019,46
2020,39
2018,18
2016,15
2017,11
2015,5
2014,4


 Maximum accident has occured dureing with years sugegsts possibly use of tesla car increased so accidents are also increased

In [26]:
# How many crashes happen over time?
events_per_date = Tesla_df_clean.groupby('Date').size()
events_per_date

,0
Date,
2013-04-02,1
2013-11-02,1
2014-07-04,2
2014-07-14,1
2014-12-30,1
...,...
2022-12-18,1
2022-12-19,1
2022-12-22,1


In [27]:
# Monthwise Analysis of Accident
Tesla_df_clean['Date'].dt.month.value_counts()



,count
Date,
11,35
12,33
5,28
6,28
7,27
9,26
8,24
3,24
4,23


In the month of May, june, july, Nov, and Dec the number of accident records are high

In [28]:
# Date wise Analysis of accident
Tesla_df_clean['Date'].dt.day.value_counts()

,count
Date,
7,21
17,13
18,13
1,13
12,13
21,13
20,13
26,12
24,12


In [29]:
# Date wise Analysis of death having high number of Death records
Tesla_df_clean['Date'].dt.day.value_counts()[Tesla_df_clean['Date'].dt.day.value_counts() > 10]

,count
Date,
7,21
17,13
18,13
1,13
12,13
21,13
20,13
26,12
24,12


There is no specific trend of accident cases with the date

In [30]:
# Weekwise analysis of death
Tesla_df_clean['Date'].dt.weekday.value_counts()

,count
Date,
5,53
6,49
4,45
0,43
2,41
1,35
3,28


Accidents rates are very near the weekends having indices 5,6 over the weekdays having indices 0,1,2,3. Therefore possibly more peoples were driving car in weekends

In [75]:
Tesla_df_clean['Country'].value_counts()

,count
Country,
USA,215
China,16
Germany,11
Canada,10
Netherlands,6
UK,5
Norway,4
Holland,3
Switzerland,3


USA has siginificant records of the accidents. Possibly they are the heavy user of this car.

In [87]:
# Death percentages in USA
Tesla_df_clean[Tesla_df_clean['Country'].str.strip() == 'USA']['Deaths'].sum()/Tesla_df_clean['Deaths'].sum()*100


np.float64(73.93767705382436)

In [78]:
Tesla_df_clean['Country'].unique()

array([' USA ', ' Canada ', ' China ', ' Mexico ', ' UK ', ' Germany ',
       ' Finland ', ' Australia ', ' Netherlands ', ' Switzerland ',
       ' France ', ' Denmark ', ' Belgium ', ' Portugal ',
       ' South Korea ', ' Norway ', ' Taiwan ', ' Slovenia ', ' Austria ',
       ' Ukraine ', ' Spain ', ' Holland ', ' Japan '], dtype=object)

In [80]:
Tesla_df_clean[Tesla_df_clean['Country'].str.strip() == 'USA']['State'].value_counts()

,count
State,
CA,75
FL,24
CA,16
AZ,8
FL,8
OH,6
GA,6
NY,5
PA,5


In [84]:
Tesla_df_clean['Deaths'].sum()

np.int64(353)

In [83]:
Tesla_df_clean[Tesla_df_clean['State'].str.strip() == 'CA']['Deaths'].sum()/Tesla_df_clean['Deaths'].sum()*100

np.float64(31.1614730878187)

31% of the overall death occured in CA state only

In [31]:
Tesla_df_clean.columns

Index(['Case #', 'Date', 'Country', 'State', 'Description', 'Deaths',
       'Tesla driver', 'Tesla occupant', 'Other vehicle', 'Cyclists/ Peds',
       'TSLA+cycl / peds', 'Model', 'Autopilot claimed',
       'Verified Tesla Autopilot Deaths',
       'Verified Tesla Autopilot Deaths + All Deaths Reported to NHTSA SGO'],
      dtype='object')

In [32]:
# Checkthe accident involved differnt numbers of deaths
Tesla_df_clean['Deaths'].value_counts()

,count
Deaths,
1,247
2,38
3,6
4,3


All the accident involed one or many death cases. Fortunately, in most of the accidents the number of death cases was 1.
a few cases observed where the number of deaths are more than 2.  

In [33]:
# Check the accident involved differnt numbers of drivers
Tesla_df_clean['Tesla driver'].value_counts()/len(Tesla_df_clean)*100

,count
Tesla driver,
-,60.204082
1,39.795918


In 60% accident cases the drivers were allived suggests somehow car safety is moderatley good

In [34]:
Tesla_df_clean['Tesla occupant'].value_counts()

,count
Tesla occupant,
-,251
1,39
2,3
3,1


There are 251 cases where no Tesla occupants were dead. 43 tesla occupants were dead according to the report

In [35]:
# Check the accident involved different numbers of drivers where a Tesla driver was involved
Tesla_df_clean[Tesla_df_clean['Tesla driver'] == '1']['Tesla occupant'].value_counts()

,count
Tesla occupant,
-,96
1,18
2,2
3,1


there are 18 cases where one Occupant was dead, 2 cases where 2 occupants were dead and 1 cases where 3 occupants were dead along with the driver

In [36]:
Tesla_df_clean['Other vehicle'].value_counts().sum()

np.int64(294)

In [37]:
Tesla_df_clean['Other vehicle'].value_counts()/Tesla_df_clean['Other vehicle'].value_counts().sum()*100

,count
Other vehicle,
-,62.585034
1,32.312925
2,3.741497
3,1.020408
4,0.340136


So, 62% of the accident does not involves hit to any other vehicle but rest of the percentages (32%) are involved with hit of some other cars possibly failure in the detection/mechanical techniques


In [39]:
Tesla_df_clean['Cyclists/ Peds'].value_counts()/Tesla_df_clean['Cyclists/ Peds'].value_counts().sum()*100

,count
Cyclists/ Peds,
-,85.034014
1,14.285714
2,0.680272


 Around 15% times Tesla car hit the pedestrian/cyclist. Possibly engine failure or vission detection problem.  

In [40]:
(pd.to_numeric(Tesla_df_clean['Tesla driver'], errors='coerce').sum()/Tesla_df_clean['Deaths'].sum())*100



np.float64(33.14447592067989)

In [41]:
(pd.to_numeric(Tesla_df_clean['Tesla occupant'], errors='coerce').sum()/Tesla_df_clean['Deaths'].sum())*100

np.float64(13.59773371104816)

33% was driver from the total death cases, 14% was the occupants Therefore, rests were either pedestrians or cyclist (~53%). It is a very terrific report about the automous car

In [51]:
tesla_occupant_numeric = pd.to_numeric(Tesla_df_clean['Tesla occupant'], errors='coerce').fillna(0)
cyclists_peds_numeric = pd.to_numeric(Tesla_df_clean['Cyclists/ Peds'], errors='coerce').fillna(0)
tesla_driver_numeric = pd.to_numeric(Tesla_df_clean['Tesla driver'], errors='coerce').fillna(0)

count = Tesla_df_clean[
    (tesla_occupant_numeric > 0) &
    (cyclists_peds_numeric > 0) &
    (tesla_driver_numeric == 1)
].shape[0]

print(count)

0


There are no such accidents where driver, tesla occupant, and cyclist/pedestrians all were involved

In [52]:
Tesla_df_clean['Model'].value_counts()

,count
Model,
-,181
S,45
3,39
X,17
Y,10
1,1
2,1


Among All the accidents record, S model has highest accident rates follwed by 3 and X. Probably, S,3, X have some safety issues.


In [58]:
Tesla_df_clean['Model'].unique()

array([' - ', ' Y ', '1', '2', '3', ' S ', ' X '], dtype=object)

In [59]:
Tesla_df_clean[Tesla_df_clean['Model']==' S ']['Deaths'].value_counts()

,count
Deaths,
1,38
2,6
3,1


In [65]:
Tesla_df_clean[Tesla_df_clean['Model']==' S ']['Cyclists/ Peds'].value_counts()

,count
Cyclists/ Peds,
-,40
1,5


In [56]:
Tesla_df_clean[Tesla_df_clean['Model']=='3']['Deaths'].value_counts()

,count
Deaths,
1,33
2,4
3,2


In [66]:
Tesla_df_clean[Tesla_df_clean['Model']=='3']['Cyclists/ Peds'].value_counts()

,count
Cyclists/ Peds,
-,33
1,6


In [60]:
Tesla_df_clean[Tesla_df_clean['Model']==' X ']['Deaths'].value_counts()

,count
Deaths,
1,15
2,2


In [63]:
Tesla_df_clean[Tesla_df_clean['Model']==' X ']['Cyclists/ Peds'].value_counts()

,count
Cyclists/ Peds,
-,13
1,3
2,1


# Peds/cyclist death rate:
1. S-Model=Deact of peds/cyclist(5)/Death(38)=13%
2. 3-Model=Deact of peds/cyclist(6)/Death(33)=18%
3. X-Model=Deact of peds/cyclist(3)/Death(15)=20% --Highest


In [71]:
Tesla_df_clean['Autopilot claimed'].value_counts ()

,count
Autopilot claimed,
-,259
1,32
2,3


In [74]:
Tesla_df_clean['Verified Tesla Autopilot Deaths'].value_counts ()

,count
Verified Tesla Autopilot Deaths,
-,277
1,13
2,3
-,1


A very less proportion of the data has verfied autopilot death claimed